### PR TITLE
feat:exposing CBRs in DA

### PR DESCRIPTION
**Pull request by Aditya-ranjan-16**
### Description

Exposed CBRs in DA 

[GIT Issue](https://github.ibm.com/GoldenEye/issues/issues/11825)

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

Exposes CBRs in DA 

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Don't forget when adding a new input variable to a DA, it needs to go into ibm_catalog.json too so it can be ordered.

Also this PR is missing supporting markdown doc which the variable descrption needs to point to. We already have one in this repo [here](https://github.com/terraform-ibm-modules/terraform-ibm-observability-da/blob/main/solutions/instances/DA-types.md), so you should update that

File: solutions/standard/DA-types.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Configuring complex inputs for KMS in IBM Cloud projects
+
+Several optional input variables in the IBM Cloud [KMS all inclusive deployable architecture](https://cloud.ibm.com/catalog#deployable_architecture) use complex object types. You specify these inputs when you configure deployable architecture.
+
+* Rules For Context-Based Restrictions (`cbr_rules`)
```
Comment (Ak-sky) [inline solutions/standard/DA-types.md:5]: Rules are for the service (KMS in this case) which is known as CBR rules, you need to update this accordingly.

File: solutions/standard/DA-types.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+
+## Rules For Context-Based Restrictions <a name="cbr_rules"></a>
+
+The `cbr_rules` input variable allows you to provide a list of rules for the target service in this case the KMS all inclusive DA instance to enforce access restrictions for the instance based on the context of access requests. Context are criterias that include the network location of access requests, the endpoint type from where the request is sent etc ..
```
Comment (Ak-sky) [inline solutions/standard/DA-types.md:10]:
```suggestion
The `cbr_rules` input variable allows you to provide a rule for the target service, in this case the KMS all inclusive DA instance, to enforce access restrictions for the instance based on the context of access requests. Context are criteria that include the network location of access requests, the endpoint type from where the request is sent, etc.
```

File: solutions/standard/DA-types.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+- Variable name: `cbr_rules`.
+- Type: A list of objects. Each object represents a set of rules
```
Comment (Ak-sky) [inline solutions/standard/DA-types.md:13]: Though this is a list but the list will only have one rule for the target service and not the set of rules.
Issue created for this - https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/364

File: solutions/standard/DA-types.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+### Options for cbr_rules
+
+  - `description` (required): The description of the rule to create.
+  - `account_id` (required): The IBM Cloud Account ID
+  - `rule_contexts` (required): (List) The contexts the rule applies to
+      - `attributes` (optional): (List) Individual context attributes
+        - `name` (required): The attribute name.
+        - `value`(required): The attribute value.
+
+  - `enforcement_mode` (required): The rule enforcement mode: `enabled` - The restrictions are enforced and reported. This is the default. `disabled` - The restrictions are disabled. Nothing is enforced or reported. `report` - The restrictions are evaluated and reported, but not enforced.
```
Comment (Ak-sky) [inline solutions/standard/DA-types.md:25]: Please make the `enforcement_mode` type as nested bullet points, this one will be confusing for the users.

**Review by Ak-sky — CHANGES_REQUESTED** [review]
Left few comments.

**Comment by Ak-sky** [comment]
@Aditya-ranjan-16, also please link the GIT issue in this PR description.

File: solutions/standard/DA-types.md (hunk 1 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-types.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/806305d97382a08396e3ba9111e5efdbda2cf2cc

File: solutions/standard/DA-types.md (hunk 2 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-types.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/commit/806305d97382a08396e3ba9111e5efdbda2cf2cc

File: solutions/standard/DA-types.md (hunk 3 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-types.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/commit/806305d97382a08396e3ba9111e5efdbda2cf2cc

File: solutions/standard/DA-types.md (hunk 4 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-types.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/commit/806305d97382a08396e3ba9111e5efdbda2cf2cc

File: solutions/standard/variables.tf [hunk 5]
```diff
@@ -115,3 +115,29 @@ variable "keys" {
...
+        value = string
+    }))) }))
+    enforcement_mode = string
+    operations = optional(list(object({
+      api_types = list(object({
+        api_type_id = string
+      }))
+    })))
+  }))
+  description = "(Optional, list) List of context-based restrictions rules to create. [Learn more](https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/tree/main/solutions/standard/DA-types.md)"
```
Comment (maheshwarishikha) [inline solutions/standard/variables.tf:139]: link is wrong, please update.

File: solutions/standard/DA-cbr_rules.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Configuring complex inputs for KMS in IBM Cloud projects
+
+Several optional input variables in the IBM Cloud [KMS all inclusive deployable architecture](https://cloud.ibm.com/catalog#deployable_architecture) use complex object types. You specify these inputs when you configure deployable architecture.
+
+* Context-Based Restrictions Rules (`cbr_rules`)
+
+
+## Rules For Context-Based Restrictions <a name="cbr_rules"></a>
+
+The `cbr_rules` input variable allows you to provide a rule for the target service, in this case the KMS all inclusive DA instance, to enforce access restrictions for the instance based on the context of access requests. Context are criteria that include the network location of access requests, the endpoint type from where the request is sent, etc.
```
Comment (maheshwarishikha) [inline solutions/standard/DA-cbr_rules.md:10]: `in this case the KMS all inclusive DA instance` - its not correct. KMS all inclusive is just the name of the repo not the DA name. Also, we should not call is as DA instance.

Can we say:
`The `cbr_rules` input variable allows you to provide a rule for the target service to enforce access restrictions for the service based on the context of access requests.`

Comment (maheshwarishikha) [reply solutions/standard/DA-cbr_rules.md:10]: `Context are criteria` --> 
either `Contexts are criteria` OR `Context is criteria`

Please check the Cloud docs to be consistent in the statements and then update

File: solutions/standard/DA-cbr_rules.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+- Variable name: `cbr_rules`.
+- Type: A list of objects. Allowes only one object representing a rule for the target service
```
Comment (maheshwarishikha) [inline solutions/standard/DA-cbr_rules.md:13]: Typo - `Allowes`

File: solutions/standard/DA-cbr_rules.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  - `enforcement_mode` (required): The rule enforcement mode can have the following values:
+      - `enabled` - The restrictions are enforced and reported. This is the default. '
+      - `disabled` - The restrictions are disabled. Nothing is enforced or reported.
+      - `report` - The restrictions are evaluated and reported, but not enforced.
+  - `operations` (optional): The operations this rule applies to
+    - `api_types`(required): (List) The API types this rule applies to.
+        - `api_type_id`(required):The API type ID
+
+### Example Rules For Context-Based Restrictions Configuration
```
Comment (maheshwarishikha) [inline solutions/standard/DA-cbr_rules.md:33]: Rules -> Rule

File: solutions/standard/DA-cbr_rules.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  account_id = "abac0df06b6......."
+  rule_contexts= [{
+      attributes = [
+                {
+                              "name" : "endpointType",
+                              "value" : "private"
+                },
+                {
+                  name  = "networkZoneId"
+                  value = "559052eb8f43302824e7........." # pragma: allowlist secret
```
Comment (maheshwarishikha) [inline solutions/standard/DA-cbr_rules.md:48]: same as above

File: solutions/standard/variables.tf [hunk 10]
```diff
@@ -115,3 +115,29 @@ variable "keys" {
...
+  default     = []
+  # Validation happens in the rule module
```
Comment (maheshwarishikha) [inline solutions/standard/variables.tf:141]: I think...we can remove this `# Validation happens in the rule module`

File: ibm_catalog.json [hunk 11]
```diff
@@ -138,6 +139,9 @@
             {
               "key": "prefix"
             },
+            {
+              "key":"cbr_rules"
+            },
```
Comment (maheshwarishikha) [inline ibm_catalog.json:144]: In this json file, ordering of variables matters. So we should place it logically. I dont think, it fits well before the instance name. I suggest to move down after keys.

File: solutions/standard/variables.tf (hunk 5 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/variables.tf]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/DA-cbr_rules.md (hunk 6 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-cbr_rules.md]: done https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/DA-cbr_rules.md (hunk 7 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-cbr_rules.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/DA-cbr_rules.md (hunk 8 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-cbr_rules.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/DA-cbr_rules.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    - `api_types`(required): (List) The API types this rule applies to.
+        - `api_type_id`(required):The API type ID
+
+### Example Rules For Context-Based Restrictions Configuration
+
+```hcl
+cbr_rules = [
+  {
+  description = "KMS can be accessed from ...."
+  account_id = "abac0df06b6......."
```
[the remark this answers is not in the document]
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-cbr_rules.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/DA-cbr_rules.md (hunk 9 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/DA-cbr_rules.md]: fixed https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: solutions/standard/variables.tf (hunk 10 above)
Comment (Aditya-ranjan-16) [reply solutions/standard/variables.tf]: done https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

File: ibm_catalog.json (hunk 11 above)
Comment (Aditya-ranjan-16) [reply ibm_catalog.json]: done https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/pull/594/commits/67a2aa1c5119d9bbaed014260a89ecfed7c12f95

**Review by Ak-sky — APPROVED** [review]

**Pull request merged**